File: Coin Change[11-August-2023]/Tabulation/Solution.cpp
//  >> C++ CODE <<

class Solution {
  public:
    long long int count(int coins[], int N, int sum) {

        vector<vector<long long int>> dp(N + 1, vector<long long int>(sum + 1, 0));

        for (int i = N; i >= 0; i--) {
            for (int j = sum; j >= 0; j--) {
                if (j == sum) dp[i][j] = 1;
                else if (i == N) dp[i][j] = 0;
                else {
                    if (j + coins[i] > sum) {
                        dp[i][j] = (i + 1 <= N) ? dp[i + 1][j] : 0;
                    } else {
                        dp[i][j] = dp[i][j + coins[i]] + ((i + 1 <= N) ? dp[i + 1][j] : 0);
                    }
                }
            }
        }

        return dp[0][0];

    }
};

File: Coin Change[11-August-2023]/Tabulation/Solution.py
#  >> PYTHON CODE <<

class Solution:
    def count(self, coins, N, Sum):
        dp=[[0 for i in range(Sum+1)] for j in range(N+1) ]
        
        for i in range(N,-1,-1):
            for j in range(Sum,-1,-1):
                if(j==Sum): dp[i][j]=1
                elif i==N: dp[i][j]=0
                else:
                    if(j+coins[i]>Sum):
                        dp[i][j]=dp[i+1][j] if (i+1)<=N else 0
                    else:
                        dp[i][j]=(dp[i][j+coins[i]]) + (dp[i+1][j] if (i+1)<=N else 0)
        
        return dp[0][0]

